chalk-solve/src/rust_ir.rs

```diff
@@ -93,15 +93,24 @@ impl<I: Interner> AdtDatum<I> {
 
 #[derive(Clone, Debug, PartialEq, Eq, Hash, Fold, HasInterner)]
 pub struct AdtDatumBound<I: Interner> {
-    pub fields: Vec<Ty<I>>,
+    pub variants: Vec<AdtDatumVariant<I>>,
     pub where_clauses: Vec<QuantifiedWhereClause<I>>,
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Fold, HasInterner)]
+pub struct AdtDatumVariant<I: Interner> {
+    pub fields: Vec<Ty<I>>,
+}
+
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub struct AdtFlags {
     pub upstream: bool,
     pub fundamental: bool,
     pub phantom_data: bool,
+    /// `true` if the ADT was lowered from a `struct` (this information is required
+    /// for determining if we have built-in `Unsize` impls as `enum`s can never be
+    /// `Unsize`).
+    pub struct_: bool,
```

Comment:
We should use an enum (as described in [this comment](https://github.com/rust-lang/chalk/pull/524/files#r439853744)) here as well.